### DISCUSSION TITLE
zimfw update finishes with non-zero return code

### DISCUSSION THREAD

**Discussion by PatTheMav** [zimfw/zimfw#459] [Q&A]
I won't create an issue for this yet, because I haven't been able to reproduce this issue consistently, but for a few days know I have `zimfw update` finish with a non-zero return code.

All my zim modules are simple git checkouts without any post-pull actions. All updates are successful (every item has a green bracket), but the function returns non-zero. I did a run with trace, and here's what I got (excuse the German-ness of git's output - "Bereits aktuell" should mean "Already up-to-date"):

```
+_zimfw_version_check:7> print 1.9.0
) utility: Bereits aktuell.
) zimfw-extras: Bereits aktuell.
) environment: Bereits aktuell.
) completion: Bereits aktuell.
) zsh-history-substring-search: Bereits aktuell.
) input: Bereits aktuell.
) zsh-autosuggestions: Bereits aktuell.
) zsh-syntax-highlighting: Bereits aktuell.
) powerlevel10k: Bereits aktuell.
) archive: Bereits aktuell.
) zsh-completions: Bereits aktuell.
+(eval):3> j=6442
+(eval):4> wait 6442
+(eval):6> case 0 (0)
+(eval):3> j=6445
+(eval):4> wait 6445
+(eval):6> case 19 (0)
+(eval):6> case 19 (<1-125>|128)
+(eval):8> ret=123 
+(eval):3> j=6447
+(eval):4> wait 6447
+(eval):6> case 0 (0)
+(eval):3> j=6449
+(eval):4> wait 6449
+(eval):6> case 0 (0)
+(eval):3> j=6451
+(eval):4> wait 6451
+(eval):6> case 0 (0)
+(eval):3> j=6455
+(eval):4> wait 6455
+(eval):6> case 0 (0)
+(eval):3> j=6456
+(eval):4> wait 6456
+(eval):6> case 0 (0)
+(eval):3> j=6462
+(eval):4> wait 6462
+(eval):6> case 0 (0)
+(eval):3> j=6465
+(eval):4> wait 6465
+(eval):6> case 0 (0)
+(eval):3> j=6470
+(eval):4> wait 6470
+(eval):6> case 0 (0)
+(eval):3> j=6473
+(eval):4> wait 6473
+(eval):6> case 0 (0)
+(eval):15> _zajobs=( ) 
+zargs:283> (( ARGC ))
+zargs:315> return 123
+zimfw:74> return 1
```

I haven't figured out where that final `zargs` iteration comes from, but if it happened at least once, all consecutive calls of `zimfw update` will fail as well.

It should be noted that I do run zimfw upgrade automatically as part of my Dotfile's update procedure via GNU makefiles:

```
ZIMFW_INIT = source $${ZDOTDIR:-$${HOME}}/.zim/init.zsh &&
update-zimfw : | $(ZIM_LOCATION)
	@$(OUTPUT) "\033[32m==> \033[37;1mUpdating zim framework...\033[0m"
	@$(ZIMFW_INIT) zimfw upgrade
	@$(ZIMFW_INIT) zimfw install
	@$(ZIMFW_INIT) zimfw update
```

That block worked without issue until a few weeks ago. @ericbn I think you added support for a post-install hook, maybe that's incompatible with how I update zimfw now?

**Comment by ericbn** [comment]
Hi @PatTheMav. Thanks for starting this discussion.

I've also observed this issue, and was able to track down a few things:
* I consistently see the issue in macOS 12.4 with the Homebrew-installed Zsh 5.9, but never with earlier versions of Homebrew-installed Zsh or the also early macOS-provided /bin/zsh (5.8.1).
* The issue seems to be zargs "randomly" returning 123 for one of the interactions, [here](https://github.com/zimfw/zimfw/blob/741bb78a04c978e223b4df845ed6ae9291e54eec/src/stage2/50_zimfw_install_update.zsh.erb#L2), even though _zimfw_run_tool is always returning 0.
* I also see the "case 19" in the xtrace output, but as far as I could check, _zimfw_run_tool was always returning 0, so no ideia where that 19 came from.
* [This](https://github.com/zsh-users/zsh/blob/545c42cdac25b73134a9577e3c0efa36d76b4091/Functions/Misc/zargs#L11) is what zargs says about returning 123, but that's not what seems to be happening.
* Even with Homebrew-installed Zsh 5.9 in in macOS 12.4, the issue never happens when I remove the -P0 from zargs, [here](https://github.com/zimfw/zimfw/blob/741bb78a04c978e223b4df845ed6ae9291e54eec/src/stage2/50_zimfw_install_update.zsh.erb#L2). 
* The zargs implementation of -P recently changed, [here](https://github.com/zsh-users/zsh/commit/67f932e7c5cfad02cdcc552dcd57e63536b7f965). 

Which OS and Zsh version are you using? Can you please try different variations of OS and Zsh available to you?

PS: the "Bereits aktuell" in your output is just lovely. :- )

**Reply by PatTheMav** [reply]
I'm on macOS 12.4 on all my machines and Zsh 5.9 via Homebrew. I do have zim running on Debian boxes with Zsh 5.8 which don't seem to exhibit the issue.

Also with Zsh 5.8 shipped with Monterey I haven't been able to reproduce this issue yet (tbf it was only 100% reproducible via the Makefile invocation for some reason).

**Reply by ericbn** [reply]
@Eriner and contributors, can you please help checking if this issue happens outside macOS with Zsh 5.9 installed via Homebrew? I think we can assume the issue does not happen with Zsh <5.9 in any OS. There's still the question whether it happens in Zsh 5.9 in other OSs...

**Reply by PatTheMav** [reply]
FWIW I'm not able to replicate it with Zsh 5.9 on my Ubuntu VM with the same Dotfile/Makefile setup.

**Reply by ericbn** [reply]
Tested with the zsh-users docker image (which is based on [bitnami/minideb:buster](https://github.com/zsh-users/zsh-docker/blob/6a24b10de7fa8eb5b785ce7ce25fc295704bd9bf/Dockerfile#L48)), and was also unable to replicate:
```
% docker run --rm -it zshusers/zsh:5.9
# apt update
# apt install -y curl git
# curl -fsSL https://raw.githubusercontent.com/zimfw/install/master/install.zsh | zsh
) Using Zsh version 5.9
) ZIM_HOME not set, using the default one.
) Changed your default shell to /usr/bin/zsh
) Downloaded the Zim script to /root/.zim/zimfw.zsh
) Prepended Zim template to /root/.zimrc
) Prepended Zim template to /root/.zshrc
) Installed modules.
All done. Enjoy your Zsh IMproved! Restart your terminal for changes to take effect.
#
```
How I replicate in my macOS:
```
% cd ${$(mktemp -d):A}
% touch .zshrc
% ZDOTDIR=${PWD} HOME=${PWD} ZIM_HOME=${PWD}/.zim exec zsh
% curl -fsSL https://raw.githubusercontent.com/zimfw/install/master/install.zsh | zsh
) Using Zsh version 5.9
) Your ZIM_HOME is the default one.
) Zsh is your default shell.
) Downloaded the Zim script to /.../.zim/zimfw.zsh
) Prepended Zim template to /.../.zimrc
) Prepended Zim template to /.../.zshrc

x Could not install modules.
%
```

**Comment by ericbn** [comment]
I've got a minimal example to reproduce this:
```
% f() { curl -Sso /dev/null 'https://example.com'; return 0 }
% zargs -n 1 -P 0 -- {1..13} -- f
% print ${?}
123
```
If the `f` function is simplified, I don't see the error. Not sure if it's related to network activity. Added the extra `return 0` to force that return code. If the `-P 0` is removed, I don't see the error. So looks like it's related to concurrency. If the range is shortened, the error will be less frequent. Even with `{1..13}` it happens to sometimes succeed in my Mid 2015 2.2 GHz Quad-Core Intel Core i7.

**Reply by ericbn** [reply]
Opened https://github.com/Homebrew/homebrew-core/issues/101937

**Comment by ericbn** [comment]
As I'm not sure there will be a fix for this so soon -- it can even potentially be a macOS bug --, I'm suppressing the zargs return value in this case for the time being. Released zimfw 1.9.1 with this patch.
